modules/autotagging-commons/src/main/java/org/opensearch/rule/action/TransportCreateRuleAction.java

```diff
@@ -0,0 +1,117 @@
+
+    /**
+     * Constructor for TransportCreateRuleAction
+     * @param transportService - a {@link TransportService} object
+     * @param actionFilters - a {@link ActionFilters} object
+     * @param threadPool - a {@link ThreadPool} object
+     * @param rulePersistenceServiceRegistry - a {@link RulePersistenceServiceRegistry} object
+     * @param ruleRoutingServiceRegistry - a {@link RuleRoutingServiceRegistry} object
+     */
+    @Inject
+    public TransportCreateRuleAction(
+        TransportService transportService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        RulePersistenceServiceRegistry rulePersistenceServiceRegistry,
+        RuleRoutingServiceRegistry ruleRoutingServiceRegistry
+    ) {
+        super(CreateRuleAction.NAME, actionFilters, transportService.getTaskManager());
+        this.ruleRoutingServiceRegistry = ruleRoutingServiceRegistry;
+        this.threadPool = threadPool;
+        this.rulePersistenceServiceRegistry = rulePersistenceServiceRegistry;
+
+        transportService.registerRequestHandler(
+            CreateRuleAction.NAME,
+            RULE_THREAD_POOL_NAME,
+            CreateRuleRequest::new,
+            new TransportRequestHandler<CreateRuleRequest>() {
+                @Override
+                public void messageReceived(CreateRuleRequest request, TransportChannel channel, Task task) {
+                    executeLocally(request, ActionListener.wrap(response -> {
+                        try {
+                            channel.sendResponse(response);
+                        } catch (IOException e) {
+                            logger.error("Failed to send CreateRuleResponse to transport channel", e);
+                            throw new TransportException("Fail to send", e);
+                        }
+                    }, exception -> {
+                        try {
+                            channel.sendResponse(exception);
+                        } catch (IOException e) {
+                            logger.error("Failed to send exception response to transport channel", e);
+                            throw new TransportException("Fail to send", e);
+                        }
+                    }));
+                }
+            }
+        );
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateRuleRequest request, ActionListener<CreateRuleResponse> listener) {
+        ruleRoutingServiceRegistry.getRuleRoutingService(request.getRule().getFeatureType())
+            .handleCreateRuleRequest(request, new ActionListener<>() {
+                @Override
+                public void onResponse(Void unused) {
+                    executeLocally(request, listener);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(new RuntimeException("Local execution of rule creation failed"));
+                }
+            }, listener);
```

Comment:
Why we do need to pass 2 listeners here? Just one should suffice.

You can create a new listener by wrapping the existing, and only sending that one. Something like below

```
ActionListener.wrap(unused -> executeLocally(request, listener),
            ex  -> listener.onFailure(ex));
```